tiledb/sm/rest/curl.cc

```diff
@@ -211,6 +255,22 @@ Status Curl::init(
     return LOG_STATUS(Status::RestError(
         "Error initializing libcurl; failed to set CURLOPT_ERRORBUFFER"));
 
+  rc = curl_easy_setopt(curl_.get(), CURLOPT_HEADERFUNCTION, OnReceiveData);
+  if (rc != CURLE_OK)
+    return LOG_STATUS(Status::RestError(
+        "Error initializing libcurl; failed to set CURLOPT_HEADERFUNCTION"));
+
+  rc = curl_easy_setopt(curl_.get(), CURLOPT_HEADER, 1);
+  if (rc != CURLE_OK)
+    return LOG_STATUS(Status::RestError(
+        "Error initializing libcurl; failed to set CURLOPT_HEADERFUNCTION"));
+
+  /* set url to fetch */
+  rc = curl_easy_setopt(curl_.get(), CURLOPT_HEADERDATA, res_headers_);
```

Comment:
Assuming you have made my other requested changes:
1. `res_headers_` is now of type `std::unordered_map<std::string, std::string>`, and
2. The `redirect_mtx_` is passed in to this `init` function

The best thing to do here is to pass in a struct to wrap the three things we want: the uri, the map, and the lock:
```
struct HeaderCbData {
  std::string uri;
  std::unordered_map<std::string, std::string> redirect_uri_map;
  std::mutex* redirect_uri_map_lock;
};
```

Instantiate an instance of the above struct, as a member variable in `Curl` and then pass the pointer here instead of `res_headers_`.